### PR TITLE
 New vault management module.

### PR DESCRIPTION
**Pull request by rjeffman**
There is a new vault management module placed in the plugins folder:

  plugins/modules/ipavault.py

The vault module allows to ensure presence and absence of vaults, manage
members and owner of the vault, and archive and retrieve vault data.

Here is the documentation for the module:

    README-vault.md

New example playbooks have been added:

    playbooks/vault/data-archive-in-asymmetric-vault.yml
    playbooks/vault/data-archive-in-symmetric-vault.yml
    playbooks/vault/data-remove-from-asymmetric-vault.yml
    playbooks/vault/data-remove-from-symmetric-vault.yml
    playbooks/vault/data-retrieve-from-asymmetric-vault.yml
    playbooks/vault/data-retrieve-from-symmetric-vault.yml
    playbooks/vault/ensure-asymetric-vault-is-absent.yml
    playbooks/vault/ensure-asymetric-vault-is-present.yml
    playbooks/vault/ensure-service-vault-is-absent.yml
    playbooks/vault/ensure-service-vault-is-present.yml
    playbooks/vault/ensure-shared-vault-is-absent.yml
    playbooks/vault/ensure-shared-vault-is-present.yml
    playbooks/vault/ensure-symetric-vault-is-absent.yml
    playbooks/vault/ensure-symetric-vault-is-present.yml
    playbooks/vault/ensure-vault-is-absent.yml
    playbooks/vault/ensure-vault-is-present.yml
    playbooks/vault/ensure-vault-member-group-is-absent.yml
    playbooks/vault/ensure-vault-member-group-is-present.yml
    playbooks/vault/ensure-vault-member-user-is-absent.yml
    playbooks/vault/ensure-vault-member-user-is-present.yml
    playbooks/vault/ensure-vault-owner-is-absent.yml
    playbooks/vault/ensure-vault-owner-is-present.yml

New tests added for the module:

    tests/vault/test_vault.yml

### CODE REVIEW

File: README-vault.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+Requirements
+------------
+
+**Controller**
+* Ansible version: 2.8+
+
+**Node**
+* Supported FreeIPA version (see above)
+
```
Comment (t-woerner) [inline README-vault.md]: KRA needs to be installed also.

File: README-vault.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+=========
+
+ipavault
+-------
+
+Variable | Description | Required
+-------- | ----------- | --------
+`ipaadmin_principal` | The admin principal is a string and defaults to `admin` | no
+`ipaadmin_password` | The admin password is a string and is required if there is no admin ticket available on the node | no
+`name` \| `cn` | The list of sudocmdgroup name strings. | no
```
Comment (t-woerner) [inline README-vault.md]: Is name optional?

File: README.md [hunk 3]
```diff
@@ -27,7 +27,7 @@ Features
 Supported FreeIPA Versions
 --------------------------
 
-FreeIPA versions 4.6 and up are supported by all roles. 
+FreeIPA versions 4.6 and up are supported by all roles.
 
 The client role supports versions 4.4 and up, the server role is working with versions 4.5 and up, the replica role is currently only working with versions 4.6 and up.
```
Comment (t-woerner) [inline README.md]: Please add vault also in Features.

File: README.md [hunk 4]
```diff
@@ -27,7 +27,7 @@ Features
 Supported FreeIPA Versions
 --------------------------
 
-FreeIPA versions 4.6 and up are supported by all roles. 
+FreeIPA versions 4.6 and up are supported by all roles.
```
Comment (t-woerner) [inline README.md]: This white space change is not related.

File: README-vault.md (hunk 1 above)
Comment (rjeffman) [reply README-vault.md]: Fixed.

File: README-vault.md (hunk 2 above)
Comment (rjeffman) [reply README-vault.md]: Fixed.

File: README.md (hunk 3 above)
Comment (rjeffman) [reply README.md]: Fixed.

**Comment by varunmylaraiah** [comment]
LGTM

File: README.md (hunk 4 above)
Comment (rjeffman) [reply README.md]: Fixed. This is something with my text editor.

File: README-vault.md [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+  become: true
+
+  tasks:
+  - ipasudocmdgroup:
+      ipaadmin_password: MyPassword123
+      name: symvault
+      username: admin
+      users: user01
+```
+`action` controls if the vault, data, member or owner will be handled. To add or remove members, set `action` to `member`. To add or remove owners, set `action` to `owner`. To add or remove data to the vault, set `action` to `data`.
```
Comment (t-woerner) [inline README-vault.md]: Why not using an owner parameter for owners? There is no need to add another action for this. Also `action: data` is not needed as `vault_data` is used at the same time.

Comment (rjeffman) [reply README-vault.md]: The use of `action: data` is for retrieving data stored in the vault, when `vault_data` is not present.

I agree with the usage of `owner` and `ownergruops`, which would make `action: owner` irrelevant.

Comment (rjeffman) [reply README-vault.md]: BTW, I'm not 'sold' in using `action: data`, that's what I came up with, but feel it could be better.

File: README-vault.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+Requirements
+------------
+
+**Controller**
+* Ansible version: 2.8+
+
+**Node**
+* Supported FreeIPA version (see above)
+* KRA service must be enable
```
Comment (t-woerner) [inline README-vault.md]: There is a missing "d"

File: README-vault.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+Example playbook to make sure that a vault and its members are present:
+
+```yaml
+---
+- name: Playbook to handle vaults
+  hosts: ipaserver
+  become: true
+
+  tasks:
+  - ipasudocmdgroup:
```
Comment (t-woerner) [inline README-vault.md]: This is about ipasudocmdgroup..

File: README-vault.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+      users: user01
+```
+
+`action` controls if the vault, data, member or owner will be handled. To add or remove members or vault data, set `action` to `member`.
+
+Example playbook to make sure that a vault member is present in vault:
+
+```yaml
+---
+- name: Playbook to handle sudocmdgroups
```
Comment (t-woerner) [inline README-vault.md]: Task name references sudocmdgroups

File: README-vault.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+      username: admin
+      users: user01
+      action: member
+```
+
+Example playbook to make sure that a vault owner is absent in vault:
+
+```yaml
+---
+- name: Playbook to handle sudocmdgroups
```
Comment (t-woerner) [inline README-vault.md]: ...

File: README-vault.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+      owner: user01
+      action: member
+      state: absent
+```
+
+Example playbook to make sure vault data is present in a symmetric vault:
+
+```yaml
+---
+- name: Playbook to handle sudocmdgroups
```
Comment (t-woerner) [inline README-vault.md]: ...

File: README-vault.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+        Data archived.
+        More data archived.
+      action: member
+```
+
+Example playbook to make sure vault data is absent in a symmetric vault:
+
+```yaml
+---
+- name: Playbook to handle sudocmdgroups
```
Comment (t-woerner) [inline README-vault.md]: ...

File: README-vault.md [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+      vault_password: MyVaultPassword123
+      action: member
+      state: absent
+```
+
+Example playbook to make sure vault is absent:
+
+```yaml
+---
+- name: Playbook to handle sudocmdgroups
```
Comment (t-woerner) [inline README-vault.md]: ...

Comment (rjeffman) [reply README-vault.md]: Sorry for that. Forgot to apply a stashed change.

File: README-vault.md (hunk 11 above)
Comment (rjeffman) [reply README-vault.md]: fixed.

File: README-vault.md (hunk 10 above)
Comment (rjeffman) [reply README-vault.md]: fixed

Comment (rjeffman) [reply README-vault.md]: fixed.

File: README-vault.md (hunk 12 above)
Comment (rjeffman) [reply README-vault.md]: Fixed.

File: README-vault.md (hunk 9 above)
Comment (rjeffman) [reply README-vault.md]: fixed.

File: README-vault.md (hunk 8 above)
Comment (rjeffman) [reply README-vault.md]: fixed.

File: README-vault.md (hunk 7 above)
Comment (rjeffman) [reply README-vault.md]: fixed.

File: README-vault.md (hunk 5 above)
Comment (rjeffman) [reply README-vault.md]: Modified it by droping data retrieval.

File: tests/vault/test_vault.yml [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+    ipavault:
+      ipaadmin_password: MyPassword123
+      name: symvault
+      username: user01
+      vault_password: MyVaultPassword123
+      vault_type: symmetric
+    register: result
+    failed_when: result.changed
+
+  - name: Add data to symmetric vault.
```
[the remark this answers is not in the document]
Comment (rjeffman) [reply tests/vault/test_vault.yml]: fixed
